capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+            s0 ^= self.right_rotate(message_schedule[i - 15], 18);
+            s0 ^= message_schedule[i - 15] >> 3;
+            let mut s1 = self.right_rotate(message_schedule[i - 2], 17);
+            s1 ^= self.right_rotate(message_schedule[i - 2], 19);
+            s1 ^= message_schedule[i - 2] >> 10;
+            message_schedule[i] = message_schedule[i - 16] + s0 + message_schedule[i - 7] + s1;
+        }
+
+        // Compression
+        let mut hashes = self.hash_values.get();
+        for i in 0..64 {
+            let s1 = self.right_rotate(hashes[4], 6)
+                ^ self.right_rotate(hashes[4], 11)
+                ^ self.right_rotate(hashes[4], 25);
+            let ch = (hashes[4] & hashes[5]) ^ ((!hashes[4]) & hashes[6]);
+            let constant = ROUND_CONSTANTS[i];
+            let temp1 = hashes[7] + s1 + ch + constant + message_schedule[i];
+            let s0 = self.right_rotate(hashes[0], 2)
+                ^ self.right_rotate(hashes[0], 13)
+                ^ self.right_rotate(hashes[0], 22);
+            let maj = (hashes[0] & hashes[1]) ^ (hashes[0] & hashes[2]) ^ (hashes[1] & hashes[2]);
+            let temp2 = s0 + maj;
+
+            hashes[7] = hashes[6];
+            hashes[6] = hashes[5];
+            hashes[5] = hashes[4];
+            hashes[4] = hashes[3].wrapping_add(temp1);
+            hashes[3] = hashes[2];
+            hashes[2] = hashes[1];
+            hashes[1] = hashes[0];
+            hashes[0] = temp1.wrapping_add(temp2);
+        }
+
+        let mut new_hashes = self.hash_values.get();
+        for i in 0..8 {
+            new_hashes[i] = new_hashes[i].wrapping_add(hashes[i]);
+        }
+        self.hash_values.set(new_hashes);
+    }
+}
+
+impl<'a> DigestData<'a, 32> for Sha256Software<'a> {
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, data.take()))
+        } else {
+            self.state.set(State::Data);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, data.take()))
+            } else {
+                let len = data.len();
+                self.handle.map(|handle| {
+                    self.deferred_caller.set(*handle);
+                    self.input_data.set(data);
+                    self.compute_sha256();
+                });
+                Ok(len)
+            }
+        }
+    }
+
+    fn clear_data(&self) {
+        let _ = self.initialize();
+    }
+}
+
+impl<'a> DigestHash<'a, 32> for Sha256Software<'a> {
+    fn run(
+        &'a self,
+        digest: &'static mut [u8; 32],
+    ) -> Result<(), (ErrorCode, &'static mut [u8; 32])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, digest))
+        } else {
+            self.state.set(State::Hash);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, digest))
```

Comment:
Same here, won't the state be forever `Hash`

Reply:
With the clear_data semantics of never-fail (and an Err(ErrorCode::CANCEL)), this state machine is now different. The `call` deferred call was setting state to `Idle`.